KAFKA-4541: Support for delegation token mechanism

omkreddy opened the pull request:
- Add capability to create delegation token
- Add authentication based on delegation token.
- Add capability to renew/expire delegation tokens.
- Add units tests and integration tests

omkreddy commented:
This is PR is a working version of delegation token implementation.

KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-48+Delegation+token+support+for+Kafka

work done:
1. capability to create/renew/expire delegation tokens
2. zookeeper based token Storage and token caching
3. token Based authentication using scram.
4. command-line to manage tokens. This will be migrated to new Admin API ( will raise a KIP)

Minor deviations from the KIP:
1.  tokenId field changed to string from int.
2. added owner, maxtime fields to CreateTokenResponse. This change is to make the response format similar to DescribeResponse
3. In KIP, We proposed to store scram credentials to ZK. But this is not required. Since server can create token hmac anytime, scram credentials can be created online during token creation and token retrieval from ZK. 

Ongoing Work:
1. Support for token management in Admin API (will raise a KIP)
2. System test

@junrao @harshach @ijuma @rajinisivaram 
Pls take a look whenever you get a chance.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6518/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6504/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6569/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6554/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6585/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6570/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6597/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6582/
Test PASSed (JDK 8 and Scala 2.12).

omkreddy commented:
@junrao @rajinisivaram  it will be great if you can give your initial thoughts on implementation and approach.

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -1880,6 +1880,73 @@
+    //Token Create request
+    public static final Schema TOKEN_CREATE_REQUEST_V0 = new Schema(
+            new Field("renewers", new ArrayOf(STRING),
+                    "An array of token renewers. Renewer is an Kafka PrincipalType+name string, who is allowed to renew this token before the max lifetime expires."),
```
Should PrincipalType+name be in one field or two separate fields?

omkreddy replied:
we can add two separate fields. updated the PR.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1942,6 +1949,125 @@ class KafkaApis(val requestChannel: RequestChannel,
+    }
+
+    val isTokenAuthenticated = request.session.principal.tokenAuthenticated
+    if (request.context.securityProtocol == SecurityProtocol.PLAINTEXT || isTokenAuthenticated)
```
Hmm, SSL can be used for just one-way authentication. In that case, the client is not authenticated by the server. So, we probably want to disallow token creation?

omkreddy replied:
Ok..I have excluded SSL authentication also. we can review and enable if required in future

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1942,6 +1949,125 @@ class KafkaApis(val requestChannel: RequestChannel,
+
+    val isTokenAuthenticated = request.session.principal.tokenAuthenticated
+    if (request.context.securityProtocol == SecurityProtocol.PLAINTEXT || isTokenAuthenticated)
+      sendResponseCallback( Errors.TOKEN_REQUEST_NOT_ALLOWED, -1)
```
It will be useful to define a meaningful constant for -1.

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+  def startup() = {
+    if (config.tokenAuthEnabled) {
+      zkUtils.makeSurePersistentPathExists(ZkUtils.TokenInfoPath)
+      loadCache
```
ZkNodeChangeNotificationListener already has the logic to read all existing data from ZK during init(). So, calling loadCache() seems unnecessary here.

omkreddy replied:
ZkNodeChangeNotificationListener only loads the change notifications. So we need loadCache to  build tokenCache

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+
+  import TokenManager._
+
+  type createResponseCallback = CreateTokenResult => Unit
```
Capitalize the type name?

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1942,6 +1949,125 @@ class KafkaApis(val requestChannel: RequestChannel,
+    else {
+      tokenManager.createToken(
+        request.session.principal,
+        createTokenRequest.renewers().asScala.toList,
```
We probably want to error out if the PrincipalType is not user.

omkreddy replied:
added a  check  to throw error for unwanted PrincipalTypes

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+    } else {
+
+      //Can we use UUID here?
+      val sequenceNum: Integer = zkUtils.getSequenceId(ZkUtils.TokenSequenceIdPath)
```
Perhaps using UUID is better since it makes it harder for anyone to guess a tokenId? It's also simpler since it doesn't need ZK access.

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+      val tokenId = TokenSequenceIdPrefix + sequenceNum
+
+      val issueTime = time.milliseconds
+      val maxLifeTime = if (maxTimePeriodMs <= 0) tokenMaxLifetime else maxTimePeriodMs
```
Should we bound maxLifeTime by tokenMaxLifetime too?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+      val tokenInfo = token.tokenInfo
+      val tokenInfoMap = mutable.Map[String, Any]()
+      tokenInfoMap(OwnerKey) = QuotaId.sanitize(tokenInfo.ownerAsString)
+      tokenInfoMap(RenewersKey) = tokenInfo.renewersAsString.asScala.map(e => QuotaId.sanitize(e.toString)).mkString(",")
```
Could we represent this as an array in json?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+      tokenInfoMap(ExpiryDateKey) = tokenInfo.expiryDate
+      tokenInfoMap(TokenIdKey) = tokenId
+
+      val data =  Map(VersionKey -> CurrentVersion, TokenInfoKey -> tokenInfoMap)
```
Could we just fold the tokenInfoMap into the top level?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+
+      try {
+        val token = new DelegationToken(tokenInfo, password)
+        updateToken(tokenId, token)
```
Should we hold the lock when updating the token?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+            val now = time.milliseconds
+            val tokenInfo =  token.tokenInfo
+
+            if (!allowed(principal, tokenInfo)) {
```
allowed => allowedToRenew ?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+            } else {
+              val renewTimePeriod = if (renewTimePeriodMs < 0) tokenRenewInterval else renewTimePeriodMs
+              val renewTimeStamp = now + renewTimePeriod
+              val expiryTime = Math.min(tokenInfo.maxDate, renewTimeStamp)
```
Interesting. This means that changing config.delegationTokenMaxLifeMs won't affect existing tokens. Will this be expected?

soenkeliebau replied:
I think both cases could be argued here. On the one hand it has the potential for "fraud", if for example a disgruntled employee leaves the company, briefly changes config.delegationTokenMaxLifeMs to something very large and creates a token for him to use later on..
On the other hand, having changes to this parameter effect existing tickets has the potential to create a bit of havoc on large deployments when all of a sudden tokens stop working that the users planned to get recreated sometime down the line. I've seen a lot of production issues caused by expired ssl certificates - and the expiration date on those is fixed. 

I think my vote would be to not change token expiration dates when updating delegationTokenMaxLifeMs, since the end dates will be different based on creation time anyway. Perhaps it is an option to create an admin tool to display tokens that exceed the currently configured maxtime?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+      val sequenceNum: Integer = zkUtils.getSequenceId(ZkUtils.TokenSequenceIdPath)
+      val tokenId = TokenSequenceIdPrefix + sequenceNum
+
+      val issueTime = time.milliseconds
```
We probably want to be consistent with the naming between an absolute time and a length. So, we could do issueTime/maxLifePeriod or issueTimestamp/maxLifeTime, or some other more intuitive alternatives.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/token/TokenInformation.java:
```diff
@@ -0,0 +1,140 @@
+        return issueDate;
+    }
+
+    public void setIssueDate(long issueDate) {
```
Instead of using Date, it seems that it's better to use Time?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+import scala.collection.mutable
+
+object TokenManager {
+  val DefaultHmacAlgorithm = "HmacSHA1"
```
To be more secure, it seems that this should be HmacSHA256 or HmacSHA512?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+      val tokenInfo = token.tokenInfo
+      val tokenInfoMap = mutable.Map[String, Any]()
+      tokenInfoMap(OwnerKey) = Sanitizer.sanitize(tokenInfo.ownerAsString)
+      tokenInfoMap(RenewersKey) = tokenInfo.renewersAsString.asScala.map(e => Sanitizer.sanitize(e.toString)).mkString(",")
```
e.toString can just be e. Also, SSL user name may contain comma. Perhaps, it's better to use another character such as ; as the separator?

omkreddy replied:
Ok..for now I have used ; as separator

junrao commented on clients/src/main/java/org/apache/kafka/common/security/token/TokenCache.java:
```diff
@@ -0,0 +1,112 @@
+        //Cache to hold all the tokens
+        tokenCache = credentialCache.createCache(TOKEN_CACHE, TokenInformation.class);
+        //Cache to hold tokenId->hmac mapping. This is required for renew, expire requests
+        hmacIDCache = credentialCache.createCache(TOKEN_HMAC_ID_CACHE, String.class);
```
CredentialCache is really used for caching different Scram mechanisms? So, it's a bit weird to reuse that for storing tokens and hmacID? Could we just create two specific HashMap, one for token and the other for hmacID?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+    //Update Scram Credentials
+    val scramConfig = new Properties
+    prepareScramCredentials(base64Pwd, scramConfig)
+    tokenCache.updateCredentials(tokenInfo.tokenId, scramConfig)
```
It's a bit weird for scramConfig to use Properties here. Could we just use a HashMap?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,528 @@
+    tokenCache.updateCredentials(tokenInfo.tokenId, scramConfig)
+
+    //Update hmac-id cache
+    tokenCache.addHmacIdMapping(base64Pwd, tokenInfo.tokenId)
```
Is it possible to move all the three updates to inside TokenCache?

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/RenewTokenRequest.java:
```diff
@@ -0,0 +1,113 @@
+
+    public static RenewTokenRequest parse(ByteBuffer buffer, short version) {
+        return new RenewTokenRequest(ApiKeys.RENEW_TOKEN.parseRequest(version, buffer),
+            (short) version);
```
(short) is unnecessary.

junrao commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -38,6 +38,7 @@ object ConfigType {
   val Client = "clients"
   val User = "users"
   val Broker = "brokers"
+  val Token = "tokens"
```
Token doesn't seem to used. Is anything related to token configurable?

junrao reviewed (commented):
@omkreddy : Thanks for the patch. Some comments below. Also, (1) could you add the new admin methods to org.apache.kafka.clients.admin.AdminClient since we will be phasing out kafka.admin.AdminClient? (2) could you add all the new ZK accesses to use the new ZookeeperClient?

omkreddy committed:
KAFKA-4541: Add ability to create/renew/expire/describe delegation tokens
    KAFKA-4542: DelegationToken based Authentication using SCRAM

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

omkreddy commented:
@junrao  Thanks for the review. Updated the PR with review comments. Updated the code to use KafkaZkClient. Since Java AdminClient changes requires KIP, I was planning to do this in separate PR.  Pls check.



cc  @soenkeliebau

asfgit commented:
FAILURE
 7993 tests run, 5 skipped, 3 failed.
--none--

asfgit commented:
FAILURE
 7993 tests run, 5 skipped, 1 failed.
--none--

omkreddy commented:
@junrao pinging for review.

junrao commented:
@omkreddy : I am a bit busy this week. Will review next week.

omkreddy commented:
@junrao pinging for review

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1985,6 +1993,137 @@ class KafkaApis(val requestChannel: RequestChannel,
+  def allowTokenRequests(request: RequestChannel.Request) : Boolean = {
+    if (request.session.principal.tokenAuthenticated ||
+      request.context.securityProtocol == SecurityProtocol.PLAINTEXT ||
+      request.context.securityProtocol == SecurityProtocol.SSL)
```
It's actually useful to support 2-way SSL since configuring client side certificates in a shared environment such as Storm is still painful. Actually, in the case of 1-way SSL, the authenticated user will be ANONYMOUS. So, we can simply remove this check and allow tokens to be created for ANONYMOUS user.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/CreateTokenResponse.java:
```diff
@@ -0,0 +1,167 @@
+
+    private static final Schema TOKEN_CREATE_RESPONSE_V0 = new Schema(
+        ERROR_CODE,
+        new Field("owner", STRING, "Token owner."),
```
To be consistent, this probably should be 2 separate fields of type and name like what's in CreateTokenRequest. Actually, is there a need to return owner? The client should know itself.

omkreddy replied:
I was thinking it will be easy to construct DelegationToken object from response. I am not able to find easy way to get user principal on client side.  For now, added 2 separate fields for owner field.

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,510 @@
+  }
+
+  val tokenMaxLifetime: Long = config.delegationTokenMaxLifeMs
+  val tokenRenewInterval: Long = config.delegationTokenExpiryTimeMs
```
Sometimes we use time and some other times we use interval. It would be useful to make them consistent throughout the patch.

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,510 @@
+        val tokenInfo = new TokenInformation(tokenId, owner, renewers.asJava)
+        tokenInfo.setIssueTimestamp(issueTimeStamp)
+        tokenInfo.setMaxTimestamp(maxLifeTimeStamp)
+        tokenInfo.setExpiryTimestamp(expiryTimeStamp)
```
Should we just pass in those 3 timestamps through the constructor?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,510 @@
+        } catch {
+          case e: Throwable =>
+            error("Exception while creating token", e)
+            responseCallback(CreateTokenResult(-1, -1, -1, "", Array[Byte](), Errors.forException(e)))
```
Is this for handling unexpected errors? If so, they are already handled in KafkaApis.handle().

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -1078,6 +1079,78 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
+      retryRequestUntilConnected(createRequest)
+    }
+
+    val reassignmentData = TokenInfoZNode.encode(token)
```
reassignmentData seems to be a wrong name.

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -1078,6 +1079,78 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
+    setDataResponse.resultCode match {
+      case Code.NONODE =>
+        val createDataResponse = create(reassignmentData)
+        createDataResponse.resultException.foreach(e => throw e)
```
We could potentially just do "createDataResponse .maybeThrow()" consistently in this class.

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,510 @@
+  private def updateCache(token: DelegationToken): Unit = {
+    val base64Pwd = token.passwordAsBase64String
+    val scramConfig = mutable.Map[String, String]()
+    prepareScramCredentials(base64Pwd, scramConfig)
```
Perhaps prepareScramCredentials() can just return a map instead of taking a map as the input.

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,510 @@
+    tokenInfoMap(IssueTimestampKey) = tokenInfo.issueTimestamp
+    tokenInfoMap(MaxTimestampKey) = tokenInfo.maxTimestamp
+    tokenInfoMap(ExpiryTimestampKey) = tokenInfo.expiryTimestamp
+    tokenInfoMap(TokenIdKey) = tokenInfo.tokenId()
```
The KIP wiki says there is an extra field ""credentials". So this is no longer needed? If so, could we update the wiki?

omkreddy replied:
Yes, we no longer string scram credentials. These credentials are created during server startup and during token creation. These will be maintained in memory (TokenCache). Updated the wiki

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,510 @@
+    * @return
+    */
+  def createBase64Password(tokenId: String, secretKey: SecretKey) : String = {
+    val password = createPassword(tokenId, secretKey)
```
In TokenCache, we use hmac. Perhaps it's better to use hmac instead of password here for consistency?

junrao commented on clients/src/main/java/org/apache/kafka/common/security/token/TokenCache.java:
```diff
@@ -0,0 +1,122 @@
+    private CredentialCache credentialCache = new CredentialCache();
+    //Cache to hold all the tokens
+    private Map<String, TokenInformation> tokenCache = new ConcurrentHashMap<>();
+    //Cache to hold tokenId->hmac mapping. This is required for renew, expire requests
```
This is a hmac -> tokenId map, not the other way.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/token/TokenCache.java:
```diff
@@ -0,0 +1,122 @@
+                if (credentialString == null) {
+                    cache.remove(tokenId);
+                } else {
+                    cache.put(tokenId, ScramCredentialUtils.credentialFromString(credentialString));
```
If we don't need to put the credentials in ZK, perhaps it's simpler to just pass in ScramCredential instead of an encoded string and the decode it.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/CredentialCache.java:
```diff
@@ -62,6 +62,10 @@ public C remove(String username) {
             return credentials.remove(username);
         }
 
+        public void clear() {
```
I am wondering if this is an existing problem. There is no memory barrier to cacheMap. So, not sure if we completely guarantee that entries newly added to the map are immediately visible to other threads.

omkreddy replied:
not sure If i understood your question. All the caches are generated during server startup/initialization. so all the caches should be available to all the threads. 
Changes to this class are not required. reverted the changes.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1985,6 +1993,137 @@ class KafkaApis(val requestChannel: RequestChannel,
+      def eligible(token: TokenInformation) : Boolean = {
+        val include =
+        //exclude tokens which are not requested
+          if (!owners.isEmpty && owners.asScala.filter(owner => token.ownerOrRenewer(owner)).isEmpty) {
```
exists() is probably more efficient than filter().isEmpty(). Also, since eligible will be called on all tokens, perhaps it's worth saving the result of owners.asScala and reuse.

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -1078,6 +1079,78 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
+   * Creates the required zk nodes for Token storage
+   */
+  def createTokenPaths(): Unit = {
+    createRecursive(TokenAuthZNode.path, throwIfPathExists = false)
```
This seems redundant since we are recursively creating TokensZNode.path.

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -1078,6 +1079,78 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
+      case _ => throw getDataResponse.resultException.get
+    }
+  }
+  /**
```
add a newline above.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java:
```diff
@@ -214,6 +214,14 @@ public static AbstractRequest parseRequest(ApiKeys apiKey, short apiVersion, Str
                 return new SaslAuthenticateRequest(struct, apiVersion);
             case CREATE_PARTITIONS:
                 return new CreatePartitionsRequest(struct, apiVersion);
+            case CREATE_TOKEN:
```
Perhaps it's clearer to use DELEGTION_TOKEN instead of just TOKEN in API_KEYS and the request names.

junrao commented on core/src/main/scala/kafka/admin/AdminClient.scala:
```diff
@@ -333,6 +336,35 @@ class AdminClient(val time: Time,
     ConsumerGroupSummary(metadata.state, metadata.protocol, Some(consumers), coordinator)
   }
 
+  def createToken(renewers: List[KafkaPrincipal], maxTimePeriodMs: Long = -1): (Errors, DelegationToken) = {
```
Could we move the new methods to org.apache.kafka.clients.admin.KafkaAdminClient since that's the supported public interface?

omkreddy replied:
I am planning to this in next PR. I am just waiting to finalize the API in this PR.

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/CommonFields.java:
```diff
@@ -50,4 +50,8 @@
     public static final Field.Int8 OPERATION = new Field.Int8("operation", "The ACL operation");
     public static final Field.Int8 PERMISSION_TYPE = new Field.Int8("permission_type", "The ACL permission type");
+
+    public static final Field.Str PRINCIPAL_TYPE = new Field.Str("principalType", "principalType of the Kafka principal");
```
principalType => principal_type for consistency.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1985,6 +1993,137 @@ class KafkaApis(val requestChannel: RequestChannel,
+          if (!owners.isEmpty && owners.asScala.filter(owner => token.ownerOrRenewer(owner)).isEmpty) {
+            false
+            //Owners and the renewers can describe their own tokens
+          } else if (token.ownerOrRenewer(requestPrincipal)) {
```
The current convention (see handleTopicMetadataRequest()) is that if owners is null, it indicates including all tokens. If owners is empty, we just return empty.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/CreateTokenResponse.java:
```diff
@@ -0,0 +1,167 @@
+    private final long expiryTimestamp;
+    private final long maxTimestamp;
+    private final String tokenId;
+    private final ByteBuffer password;
```
The key name is hmac, but the val name is password. It would be useful to make them consistent here and in other requests/responses.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/token/TokenCache.java:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    public void clear() {
```
This method doesn't seem to be used.

junrao commented on core/src/main/scala/kafka/admin/DelegationTokenCommand.scala:
```diff
@@ -0,0 +1,214 @@
+  def expireToken(adminClient: AdminClient, opts: DelegationTokenCommandOptions) = {
+    val hmac = opts.options.valueOf(opts.hmacOpt)
+    val expiryTimePeriodMs = opts.options.valueOf(opts.expiryTimePeriodOpt).longValue()
+    println("Calling token operation with hmac :" + hmac +" , expire-time-period : "+ expiryTimePeriodMs)
```
"Calling token operation" => "Calling expire token operation"

junrao commented on bin/kafka-delegation-tokens.sh:
```diff
@@ -0,0 +1,17 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+exec $(dirname $0)/kafka-run-class.sh kafka.admin.DelegationTokenCommand "$@"
```
Could we add the same script under bin/windows too?

junrao commented on core/src/test/scala/integration/kafka/api/TokenEndToEndAuthorizationTest.scala:
```diff
@@ -0,0 +1,88 @@
+    for (_ <- 0 until producerCount)
+      producers += createNewProducer
+    for (_ <- 0 until consumerCount) {
+      consumers += createNewConsumer
```
Is there a test method that uses producers/consumers?

omkreddy replied:
Yes. TokenEndToEndAuthorizationTest.scala extends SaslEndToEndAuthorizationTest. All the tests run with token authenticated clients.

junrao commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -534,3 +562,4 @@ object ZkData {
     } else ZooDefs.Ids.OPEN_ACL_UNSAFE.asScala
   }
 }
+
```
unneeded new line

junrao commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -68,6 +68,7 @@ object ZkUtils {
   val ConfigUsersPath = s"$ConfigPath/users"
   val ProducerIdBlockPath = "/latest_producer_id_block"
 
+
```
unneeded new line

omkreddy replied:
ZKUtils changes are not required. reverted the changes.

junrao reviewed (commented):
@omkreddy : Thanks for updated patch. Made another pass and the following are some comments.

omkreddy committed:
Address Jun's review comments

omkreddy commented:
@junrao Thanks for the review. Since adding new methods to org.apache.kafka.clients.admin.KafkaAdminClient requires KIP, I planning to do it separate PR. I will raise KIP and PR by next week.  Updated the PR. Pls take a look.

junrao commented on core/src/test/scala/integration/kafka/api/TokenEndToEndAuthorizationTest.scala:
```diff
@@ -0,0 +1,88 @@
+  override def initClients(producerSecurityProps: Properties, consumerSecurityProps: Properties) = {
+  }
+
+  def initTokenClients(token: DelegationToken) = {
```
Instead of overriding initTokenClients, would it be simpler to override clientSecurityProps() and put the additionally needed properties there?

omkreddy replied:
modified the code to minimize the changes.  Pls take a look

junrao commented on core/src/test/scala/unit/kafka/utils/JaasTestUtils.scala:
```diff
@@ -72,14 +72,14 @@ object JaasTestUtils {
   case class ScramLoginModule(username: String,
                               password: String,
-                              debug: Boolean = false) extends JaasModule {
+                              debug: Boolean = false, tokenProps: Map[String, String] = Map.empty) extends JaasModule {
```
Could we put tokenProps in a new line?

junrao commented on core/src/test/scala/unit/kafka/utils/JaasTestUtils.scala:
```diff
@@ -72,14 +72,14 @@ object JaasTestUtils {
       "username" -> username,
       "password" -> password
-    )
+    ) ++ tokenProps.map { case (user, pass) => user -> pass }
```
Are the names user, pass appropriate? It seems that they are property name and value.

junrao commented on checkstyle/suppressions.xml:
```diff
@@ -51,13 +51,10 @@
-
-    <suppress checks="JavaNCSS"
-              files="AbstractRequest.java"/>
+              files="Protocol.java|AbstractRequest.java|KerberosLogin.java|WorkerSinkTaskTest.java"/>
```
Protocol.java no longer exists.

junrao commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -158,11 +158,11 @@ object TestUtils extends Logging {
     enableSaslSsl: Boolean = false,
     rackInfo: Map[Int, String] = Map(),
-    logDirCount: Int = 1): Seq[Properties] = {
+    logDirCount: Int = 1, enableToken: Boolean = false): Seq[Properties] = {
```
Could we put enableToken in a new line?

junrao commented on core/src/main/scala/kafka/server/TokenManager.scala:
```diff
@@ -0,0 +1,477 @@
+  def expireToken(principal: KafkaPrincipal,
+                  hmac: ByteBuffer,
+                  expireLifeTimeMs: Long,
+                  expireResponseCallback: RenewResponseCallback) {
```
Perhaps it's clearer to create a separate ExpireResponseCallback.

junrao commented on core/src/test/scala/unit/kafka/security/token/TokenManagerTest.scala:
```diff
@@ -0,0 +1,344 @@
+    assertEquals(Errors.DELEGATION_TOKEN_EXPIRED, error)
+  }
+
+  def verifyCreateTokenResult(expected: CreateTokenResult): Unit = {
```
Perhaps it's better to define equal() in CreateTokenResult?

junrao commented on core/src/test/scala/unit/kafka/security/token/TokenManagerTest.scala:
```diff
@@ -0,0 +1,344 @@
+    assert(expected.hmac sameElements createTokenResult.hmac)
+  }
+
+  def renewResponseCallback(ret: Errors, timeStamp: Long): Unit = {
```
Could this and a few other non-test methods be private?

junrao commented on core/src/test/scala/unit/kafka/security/token/TokenManagerTest.scala:
```diff
@@ -0,0 +1,344 @@
+    val owner1 = SecurityUtils.parseKafkaPrincipal("User:owner1")
+    val owner2 = SecurityUtils.parseKafkaPrincipal("User:owner2")
+    val owner3 = SecurityUtils.parseKafkaPrincipal("User:owner2")
+    val owner4 = SecurityUtils.parseKafkaPrincipal("User:owner2")
```
The above 2 lines should use owner3 and owner4?

junrao commented on core/src/test/scala/unit/kafka/security/token/TokenManagerTest.scala:
```diff
@@ -0,0 +1,344 @@
+      List()
+    }
+    else {
+      def eligible(token: TokenInformation): Boolean = {
```
Is it possible to avoid duplicating eligible()?

omkreddy replied:
hmm..not found elegant way to do this.  will try  to check any other approach.

junrao reviewed (commented):
@omkreddy Thanks for the latest patch. Just a few more minor comments.

omkreddy committed:
Address Jun's comments

omkreddy commented:
@junrao  Thanks for the review. Updated the PR with review comments and some cleanups. Pls check.

soenkeliebau commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1985,6 +1992,140 @@ class KafkaApis(val requestChannel: RequestChannel,
+  }
+
+  def allowTokenRequests(request: RequestChannel.Request) : Boolean = {
+    if (request.session.principal.tokenAuthenticated || request.context.securityProtocol == SecurityProtocol.PLAINTEXT)
```
Should we extend this to disallow requests on SSL listeners that don't enforce presenting a client certificate? Or rather, SSL endpoints with a user principal of ANONYMOUS  - otherwise even authenticated users could't request a token on such an endpoint.

junrao replied:
The inconsistency is that we disallow token creation on ANONYMOUS if security protocol is PLAINTEXT, but allow it if security protocol is SSL. We could probably also disable token creation if security protocol is SSL and user is ANONYMOUS.

junrao commented on core/src/main/scala/kafka/server/DelegationTokenManager.scala:
```diff
@@ -0,0 +1,492 @@
+                             tokenId: String,
+                             hmac: Array[Byte],
+                             error: Errors) {
+  override def equals(other: Any): Boolean = {
```
If we override equals(), we would want to override hashcode() as well.

junrao commented on core/src/test/scala/integration/kafka/api/DelegationTokenEndToEndAuthorizationTest.scala:
```diff
@@ -0,0 +1,94 @@
+    super.setUp()
+  }
+
+  def getDelegationToken(): DelegationToken = {
```
Could we make this private? Also, perhaps it's more intuitive to name this createDelegationToken().

junrao commented on core/src/test/scala/integration/kafka/api/DelegationTokenEndToEndAuthorizationTest.scala:
```diff
@@ -0,0 +1,94 @@
+    adminClient.close()
+
+    //wait for token to reach all the brokers
+    Thread.sleep(100)
```
This is probably not very reliable. Could we wait until the token is in every broker's token cache by poking into each server object?

junrao commented on core/src/test/scala/unit/kafka/security/token/delegation/DelegationTokenManagerTest.scala:
```diff
@@ -15,7 +15,7 @@
   */
 
-package kafka.security.token
+package unit.kafka.security.token.delegation
```
Hmm, it seems we shouldn't include unit in package name.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1985,6 +1992,140 @@ class KafkaApis(val requestChannel: RequestChannel,
+      }
+      else {
+        val owners = if (describeTokenRequest.owners == null) null else describeTokenRequest.owners.asScala
+        def eligible(token: TokenInformation) : Boolean = {
```
Could we put eligible() in Object KafkaApis so that we can use it here and in the test?

omkreddy replied:
@junrao  Thanks for the review. made few changes to avoid the duplicate code. 
Updated thePR. Pls take a look.

junrao reviewed (commented):
@omkreddy : Thanks for updated patch. A few more minor comments below.

omkreddy committed:
Address Jun's review comments

omkreddy commented:
KIP for admin client changes: https://cwiki.apache.org/confluence/display/KAFKA/KIP-249%3A+Add+Delegation+Token+Operations+to+KafkaAdminClient
PR is also ready is [here](https://github.com/omkreddy/kafka/commit/bb41d432ee2a7b732c4a6f58ab12c2316a427c8d). I planning to start discuss thread after merging this PR.

junrao commented:
@omkreddy : Thanks for the PR. LGTM.

junrao commented:
@omkreddy : Merged the patch to trunk. Be aware that the deadline for 1.1.0 KIP freeze is 01/23. So, ideally, it would be useful to get the admin client KIP in before that.

omkreddy commented:
@junrao Thanks for your detailed review.  I start discussion on admin client changes.
